sympy/algebras/dual.py

```diff
@@ -0,0 +1,316 @@
+# References :
+# https://en.wikipedia.org/wiki/Dual_number
+from sympy import conjugate, exp, ln, sympify
+from sympy.core.expr import Expr
+
+
+class Dual(Expr):
+    """Provides basic dual number operations. Dual objects can be instantiated
+    as Dual(a, b), as in (a + b*d), where d^2 = 0.
+
+    An important use of dual numbers is automatic differentiation.
+
+
+    Examples
+    ========
+
+    >>> from sympy import symbols
+    >>> from sympy.algebras.dual import Dual
+    >>> a = symbols('a')
+    >>> d = Dual(a, 1)
+    >>> d
+    a + 1*d
+    >>> f = lambda x: x**2
+    >>> f(d)
+    a**2 + 2*a*d
+
+    We recognize the term proportional to d as the derivative of the function.
+    To obtain only the derivative we select only the dual component:
+
+    >>> f(d).b
+    2*a
+
+    In order to compute second derivatives use nested dual numbers:
+
+    >>> f(Dual(Dual(a, 1), Dual(1, 0)))
+    (a**2 + 2*a*d) + (2*a + 2*d)*d
+
+    The last component is the second derivative, which is obtained as
+
+    >>> f.b.b
+    2
+
+    """
+    _op_priority = 11.0
+
+    is_commutative = True
+
+    def __new__(cls, a=0, b=0):
+        a = sympify(a)
+        b = sympify(b)
+
+        if any(i.is_commutative is False for i in [a, b]):
```

Comment:
Contrary to your claims, I don't think that you have allowed quaternions to work here

Reply:
You are right, my code and my mind weren't in sync ;). Locally I had it already though, I've cleaned up my code and pushed it now.